### PR TITLE
Issue #14502: Handling of whitespace between generic and record header (GenericWhitespaceCheck)

### LINKED ISSUE #14502: GenericWhitespaceCheck: Handling of whitespace between generic and record header
From https://github.com/checkstyle/checkstyle/pull/14497#discussion_r1493346388 (second last paragraph)

([Check Documentation](https://checkstyle.sourceforge.io/checks/whitespace/genericwhitespace.html))

**Goal of this Issue:**
- Conclude whether whitespace between generic and record header should be considered as a violation or not. 
- Based on the conclusion necessary changes can be done.

**How it works Now:**
Currently whitespace between generic and record header is not considered a violation.

**Test.java**
```java
record Test<T> () {}
//            ^
//            +--- whitespace in question
```

**config.xml**
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="GenericWhitespace"/>
  </module>
</module>
```

**Current CLI Execution:**
```
java -jar checkstyle-10.13.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

**Expected Working:** ~~To be decided~~
`[ERROR] Test.java:1:14: '>' is followed by whitespace. [GenericWhitespace]`

### Findings that may be useful:
- The syntax is kind of similar to Issue [#14344](https://github.com/checkstyle/checkstyle/issues/14344)
- [Horizontal Whitespace](https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace) from google style, which includes the *only* places where occurrence of whitespace is valid does not includes this usage.
- In [oracle docs](https://docs.oracle.com/en/java/javase/17/language/records.html#GUID-6699E26F-4A9B-4393-A08B-1E47D4B2D263) for record classes [see first point of **Features of Record Classes**], a whitespace *is* present between generic and record header in the example.
- This check adheres to the "typical convention".

**Typical Usages of Generic Records (Anecdotal):**
- Generic records in Checkstyle, [default projects](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/projects-to-test-on-for-github-action.properties) for regression ,and those I have encountered so far, do not contain whitespace between generic and record header.
This can be verified by running the following two grep cmds:
  1. `grep -P -r "record [a-zA-Z0-9\s$_]+\<.*\>\(.*\)" dir/ | tee >(wc -l)`
  2. `grep -P -r "record [a-zA-Z0-9\s$_]+\<.*\>\s+\(.*\)" dir/ | tee >(wc -l)`
- The pattern will match for the generic records, first one matching no-whitespace between generic and record header, while the second one matching atleast a single-whitespace between generic and record header.
- Output for the cmd on Checkstyle repo: **i.** `32`; **ii:** `0`
- Output for the cmd on regression repos: **i.** `60`; **ii.** `0`
**Side note:** Arguably better way to verify that there are no such cases would be implement changes which detects space between generic and record header as a violation, and then check for differences in the report. (lmk if that is needed)

### PR DESCRIPTION
**Pull request by sktpy**
Resolves #14502 

Reasoning for the changes are mentioned in the comments.

Old Doc([site](https://checkstyle.org/checks/whitespace/genericwhitespace.html)), New Doc([site](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/05dd30d_2024155912/checks/whitespace/genericwhitespace.html#Description))
<hr>

#### Regression: ([Report](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/e3fa404_2024081446/reports/diff/index.html))

Differences found (expected)
- **Added Violations (4)**: Example (rest occurrences are similar) ([link](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/e3fa404_2024081446/reports/diff/checkstyle/xref/home/runner/work/checkstyle/checkstyle/.ci-temp/contribution/checkstyle-tester/repositories/checkstyle/src/it/resources-noncompilable/org/checkstyle/suppressionxpathfilter/recordtypeparametername/SuppressionXpathRegressionRecordTypeParameterName1.java.html#L7))
This is intended behaviour of the fix, whitespace between generic and record header is counted as violation.

Diff Regression config: https://gist.githubusercontent.com/sktpy/44a31ea4e946a0755fa3e331fd65de98/raw/bc9e57764a6ef5b9f4aa27bcb6f55115460423bb/check.xml

### CODE REVIEW

**Comment by sktpy** [comment]
GitHub, generate report

**Comment by sktpy** [comment]
GitHub, generate website

File: src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/Example1.txt [hunk 1]
```diff
@@ -23,4 +23,6 @@ List<T> list = ImmutableList.Builder<T>::new;
 sort(list, Comparable::<String>compareTo);
 // Constructor call
 MyClass obj = new <String>MyClass();
+// Record header
+record License<T>() {}
```
Comment (sktpy) [inline src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/genericwhitespace/Example1.txt:27]: Have kept the examples in the non-enabled way as GenericWhitespaceExample is being updated in the PR [#1464](https://github.com/checkstyle/checkstyle/pull/14364), if that happens to get merged before this one does, I will enable the examples else it can be updated in that PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java [hunk 2]
```diff
@@ -51,7 +51,8 @@
  * <ul>
  * <li> should not be preceded with whitespace in all cases.</li>
  * <li> should be followed with whitespace in almost all cases,
- *   except diamond operators and when preceding method name or constructor.</li></ul>
+ *   except diamond operators and when preceding method name, constructor or record header.</li>
```
Comment (sktpy) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java:54]: All similar changes signify intended changes of this PR.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java [hunk 3]
```diff
@@ -370,10 +389,9 @@ private static boolean containsWhitespaceBefore(int before, int... line) {
      * @return checks if given character is valid
      */
     private static boolean isCharacterValidAfterGenericEnd(char charAfter) {
-        return charAfter == '(' || charAfter == ')'
-            || charAfter == ',' || charAfter == '['
-            || charAfter == '.' || charAfter == ':'
-            || charAfter == ';'
+        return charAfter == ')' || charAfter == ','
```
Comment (sktpy) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java:392]: #### Removal of `charAfter == '('`:
(As determined in https://github.com/checkstyle/checkstyle/pull/14497#discussion_r1493346388 [last paragraph])

**Purpose of removal:**
To dispose mutant (replacing `charAfter == '('` with `false`) which was able to survive due to the changes in this fix in the following line:
https://github.com/checkstyle/checkstyle/blob/84eaed72081302e94be0ef096d56bddc19bebc7f/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java#L373
<hr>

**Why the mutant survives now:**
All the valid occurrences where Generic-End(`>`) can be followed by LPAREN(`(`) are the cases where it needs to be checked that such generic is not followed by whitespace, and as of this fix we have accounted for all such occurrences, making every possible input where `>` can be followed by `(` filter through `if` block of `processSingleGeneric()`:
https://github.com/checkstyle/checkstyle/blob/e3fa404dacb70d35ff507fce573698f63e36d9f6/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java#L223-L225
Hence never reaching the `else if` block which calls `isCharacterValidAfterGenericEnd()`, which makes it possible for the mutant i.e. (replacing `charAfter == '('` with false) survive all the tests.
<hr>

**Why the mutant didn't survived before:**
Because previously Check didn't counted whitespace between generic and record header as a violation, making this the killer input where `>` could be preceded by `)` which wasn't checked for succeeding whitespace, which made it produced `illegal follow` violation, hence killing the mutant.
<hr>

**Why this mutant cannot be killed:**
To kill this mutant we would need an input where `>` is followed by `(` but isn't from ones that we have already accounted for(method, ctor, record). Regression is unable to find such input, and to the best of my knowledge, such input does not exist, but if anyone is aware of such input please do tell.
<hr>

In conclusion, it is safe to remove that operand (also provides us with nano performance gain) unless anyone is aware of the killer input or if there is any plan to use this exact method in the future where that operand would be required.
The only usage of the method with mutation is in the GenericWhitespaceCheck. (1 Usage)
GenericWhitespaceCheck does not affects any other class.

**Comment by sktpy** [comment]
Failing check in CI is unrelated. ([Cirrus](https://github.com/checkstyle/checkstyle/pull/14529/checks?check_run_id=21895679122))

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java (hunk 2 above)
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/GenericWhitespaceCheck.java:54]:
```suggestion
 *   except diamond operators and when preceding a method name, constructor, or record header.</li>
```

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Excellent work! Thanks for report and website generation ahead of time to ease review.

One minor item:

**Comment by sktpy** [comment]
GitHub, generate website

**Review by nrmancuso — APPROVED** [review]
I am good as CI passes

**Comment by sktpy** [comment]
I think you may have forgot to request a review here from romani.

**Comment by sktpy** [comment]
@romani 🏓 ping pong 
pinging as I believe this might have been missed.

**Comment by sktpy** [comment]
🌵🏜️🐪
image

**Review by romani — APPROVED** [review]
Ok to merge, thanks a lot
